RFC: conditional  `enabled` field (#3066)
opentofu/opentofu: rfc/2025-07-22-conditional-enable.md

# Conditional field for enabling resources and modules

## Current status

Issue: https://github.com/opentofu/opentofu/issues/1306

Right now, OpenTofu supports conditional enable/disable of resources by using a workaround with `count`.
This approach brings a few problems, like adding indexes to resources that would be a single instance, making it harder to manage using these indexes.

This RFC proposes a new way to do a cleaner and semantic way by using a new field on the `lifecycle` block called `enabled`:

```
resource "aws_instance" "example" {
  # ...

  lifecycle {
    enabled = var.enable_server
  }
}
```

This is available not only to resources, but to modules too:

```
module "modcall" {
  source = "./mod"

  lifecycle {
    enabled = var.env == "production"
  }
}
```

## Proposed Solution

> 1. Raise errors if a resource is used while being disabled;

Reviewer: ```suggestion
1. Raise errors if a resource is used while being disabled and the dependant resources are not disabled too;
```
Right?

1. Support for conditional enabling on:

>   1. Resources - Add a new field on the `lifecycle` block called `enabled`;

Reviewer 2: Why not data sources?

Reviewer 3: `resource`, `data`, and `ephemeral` blocks all declare "resources" really. It's just an annoyance of history that the first kind of resource -- managed resources -- squatted on the generic block type name before anyone realized that there would be resources of other kinds. 

So this item is talking about resources of all kinds, just as with `count` and `for_each`. The same code handles all three when it comes to these meta-arguments.

(I don't intend this comment to mean that it would not still be worth changing the RFC texr to make this clearer. I suggest we directly name all of the block types that this new argument can appear in, to make it precise: `resource`, `data`, `ephemeral` and `module`.)

  1. Modules - Add a new `lifecycle` block for Modules but only supporting the `enabled` field.

## Technical Approach

### `lifecycle` support on Modules

Lifecycle block is currently not supported on `Modules`, but in our docs we're mentioning to users that it's reserved for future usage:

> OpenTofu does not use the lifecycle argument. However, the lifecycle block is reserved for future versions.

https://opentofu.org/docs/language/modules/syntax/#meta-arguments

It would be the right timing to start to support it, but only for the `enabled` field.

> ### Migration from `count` to `enabled`

Reviewer: I might miss some things here, but shouldn't `for_each` be handled similarly?

Reviewer 3: This section is describing some pre-existing behavior where OpenTofu notices a situation where e.g. the desired state includes `aws_instance.foo` while the previous state includes `aws_instance.foo[0]`, and so it behaves as if the author had written a `moved` block like this without requiring the author to explicitly write it:

```hcl
moved {
  from = aws_instance.foo[0]
  to   = aws_instance.foo
}
```

(there's a similar rule for migrating in the opposite direction, too.)

This rule relies on the assumption that count element zero is the one that should be kept, and all of the others discarded, creating a similar effect to what would happen if you previously had `count = 3` and now you have `count = 1` -- the "zeroth" element would be kept and the others would be planned for destroy.

There is no similar automatic rule for `for_each` because there isn't any obvious rule for choosing which of the possibly-many string instance keys is the one to keep. In that case we require the author to write out an explicit `moved` block describing which instance should be kept:

```hcl
moved {
  from = aws_instance.foo["example"]
  to   = aws_instance.foo
}
```

I'd also note that this "implied `moved` block" behavior only applies to adding/removing `count` from resources. There is no similar migration for module calls because there are many more possible ways to resolve a change like that and so OpenTofu wants the author to be explicit about what they intended.

Maybe we should add some more detail like what I described above to the RFC text so that there's some additional context about what would be handled automatically and what would not be, and we can also mention that `moved` blocks are the intended solution for the more ambiguous cases that OpenTofu doesn't handle automatically.

Reviewer 3: The only case where OpenTofu does anything automatic today is switching from nothing to `count`, or the reverse of that, on resources.

Switching from `enabled = true` to `count` > 0, and the reverse of that, is effectively the same as what currently works since it causes the same instance keys to appear, so I don't think we need any extra cider to merge that work. 

Switching from nothing to `enabled = true` and the reverse is effectively no change, because those two situations are equivalent. No address changes are required at all.

Everything else is "ambiguous" in the way I meant that, and so is not handled automatically. Adding `enabled` doesn't change anything about the other cases.

Luckily, this support would be in-place already. When using a `count` for this conditional dependency, if
the resource is changed to use `enabled`, a `move` will be done on the apply phase to remove the index and
turn into a single instance.

```
OpenTofu will perform the following actions:

  # null_resource.example[0] has moved to null_resource.example
    resource "null_resource" "example" {
        id = "8975736315378968412"
    }

Plan: 0 to add, 0 to change, 0 to destroy.
```

> ### Usage of `enabled` together with `for_each` and `count`.

Reviewer 2: I think it would be reasonable to have both, if the enabled flag has a higher priority than count or for_each. That is if it is disabled, it isn't created regardless of the value of count or for_each.

Reviewer 3: If we allow both then we need to decide what both of them together does, mechanically.

One answer is that combining `for_each` with `enabled = false` causes the same effect as `for_each = {}`, with the resource then appearing as an empty object.

Another answer is that `enabled = false` always just makes the resource be a null value of whatever type it would have been otherwise: alone it's a null object, with `for_each` a null map, and with `count` a null list.

The fact that it's not immediately obvious what it ought to mean gives me pause, though, since it means that everyone who encounters the combination will need to go learn what it means to know what effect it has. A `for_each` with a conditional or `for` expression in it has the advantage that if you already know what each of these three language features does in isolation you can infer how they will interact with each other without any additional information. 🤔

Reviewer 4: I'd just like to say that I really think we're missing out by not supporting both approaches here. And i think we should add something to the future considerations section about this.

Semantically, this feature would be used for single-instance resources or modules.

### Errors when resource is not enabled

There's some discussion about this [topic](https://github.com/opentofu/opentofu/issues/1306#issuecomment-2398120132), and the better compromise we found is that an enabled 
resource can be `null` when it's disabled, but no errors are going to be shown statically,
like when using `validate` or the language server, in an OpenTofu extension, like VSCode.
Since we need to evaluate the `enabled` expression, these errors are going to be shown dynamically,
when running `plan` or `apply`:

```
│ Error: Attempt to get attribute from null value
│
│   on main.tf line 19, in output "enabled_output":
│   19:     value = null_resource.example.id
│     ├────────────────
│     │ null_resource.example is null
│
│ This value is null, so it does not have any attributes.
```

The available data access patterns for possibly disabled resources are:

- `null_resource.example != null ? null_resource.example.id : "default value"`
- `try(null_resource.example.id, "default_value")`

> - `can(null_resource.example.id) ? null_resource.example.id : "default value"`

Reviewer 5: To be a bit more specific. Are those the *only* available patterns for accessing resources with the enabled field? Or can those be accessed regularly, and we recommend those patterns to avoid errors in case the resource is disabled?

Reviewer 3: It's true to say that there's nothing special about references to `null_resource.example` beyond the usual rules about what you can and cannot do with null values, but I do think it's worth including some examples of common patterns we're expecting so we can understand whether this design is likely to lead to something intuitive/readable.

When we finally write docs for this I think we should choose just one pattern to recommend, to encourage a consistent idiom, but we can save that decision for later.

```
resource "aws_instance" "example" {
  # ...

  lifecycle {
    enabled = var.enable_server
  }
}
```

This is available not only for resources, but for modules too:

```
module "modcall" {
  source = "./mod"

  lifecycle {
    enabled = var.env == "production"
  }
}
```

## Proposed Solution

1. Raise errors if a resource is used while being disabled and the dependent resources are not disabled too;
2. Support for conditional enabling on:
  1. Resources - Add a new field on the `lifecycle` block called `enabled`;
  2. Modules - Add a new `lifecycle` block for modules but only supporting the new `enabled` field.

## Technical Approach

### `lifecycle` support on Modules

The lifecycle block is currently not supported on `modules`, but in our docs we mention to users that it's reserved for future usage:

> OpenTofu does not use the lifecycle argument. However, the lifecycle block is reserved for future versions.

https://opentofu.org/docs/language/modules/syntax/#meta-arguments

It would be the right timing to start supporting it, but only for the `enabled` field.

## Migration from existing resources

### Migration from `count` to `enabled`

> As mentioned before, users were adding support for enabling/disabling resources by using `count = var.enabled ? 1 : 0` and this is the most common case where they want to use an `enabled` field.

Reviewer 4: Where was this mentioned before? I know you mentioned `count` in the intro but not a specific example

> Luckily, this support is [in-place already](https://github.com/opentofu/opentofu/pull/3066#discussion_r2231518392). When using a `count` for this conditional dependency, if

Reviewer 4: >Luckily, this support is [in-place already](https://github.com/opentofu/opentofu/pull/3066#discussion_r2231518392).

I feel like it took me a long time to realise that you were talking about moving resources here! a little bit of a pre-amble to each section would be appreciated to provide context of what we're talking about aside from the title.

For example, I think we should be saying something similar to: 

```
To support moving between resources where the `count = 0` workaround was used to the new `lifecycle -> enabled` approach, opentofu will automatically support an implicit move of the resource out of the box with no changes needed
```

What we have here is very jarring to read because i had no clue what "this support" is discussing and the linked comment was quite confusing.

Reviewer 3: I think we should also note that this automatic behavior exists only for resources, and that for modules (unless we change something as part of implementing this feature) it will be necessary to write an explicit `moved` block:

```hcl
moved {
  from = module.example[0]
  to   = module.example
}
```

The current behavior of automatically generating an implied `moved` block for this situation only works for resources.

    }

Plan: 0 to add, 0 to change, 0 to destroy.
```

### Migration from `for_each` to `enabled`

Unlike `count`, implicit moves are not handled by OpenTofu. This means that the user should explicitly
tell OpenTofu what needs to be moved:

```
moved {
  from = aws_s3_bucket.example["for_each_key"]  # name of the key being used by for_each
  to = aws_s3_bucket.example
}
```

So a complete example of that support would be:

```
# For_each before:

locals {
  buckets = ["bucket-1", "bucket-2"]
}

moved {
  from = aws_s3_bucket.example["bucket-1"]  # name of the key being used by for_each
  to = aws_s3_bucket.example
}

resource "aws_s3_bucket" "example" {
  for_each = var.enable ? toset(local.buckets) : []
  bucket = each.key
}

# To:
resource "aws_s3_bucket" "example" {
  lifecycle {
    enabled = var.enable
  }
}
```

It doesn't make a lot of sense in the example above to move from `for_each` to `enabled` if more than one resource is being used. `bucket-2` in the example above would be removed. This path of migration
should be used *only* if a single instance is being used and for_each is supporting that conditional.

### Usage of `enabled` together with `for_each` and `count`.

> These three different arguments would behave similarly, but with different semantics.

Reviewer 4: different how?

It could be argued that the user wants to enable a resource with 4 instances, but in our current code,

> that conditional can be changed to `count=0`.

Reviewer 4: This makes no sense to me, can you provide some more explanation here or even some examples?

> The proposal is that we do not support them together. Only one of them can be used, returning errors if we try to use the others together.

Reviewer 4: Please justify why in the RFC

Semantically, this feature would be used for single-instance resources or modules.

### Errors when resource is not enabled

There's some discussion about this [topic](https://github.com/opentofu/opentofu/issues/1306#issuecomment-2398120132), and the better compromise we found is that an enabled 
resource can be `null` when it's disabled, but no errors are going to be shown statically,

> like when using `validate` or the language server, in an OpenTofu extension, like VSCode.

Reviewer 4: ```suggestion
like when using `tofu validate` or the `tofu-ls` language server
```

Since we need to evaluate the `enabled` expression, these errors are going to be shown dynamically,
when running `plan` or `apply`:

```
│ Error: Attempt to get attribute from null value
│
│   on main.tf line 19, in output "enabled_output":
│   19:     value = null_resource.example.id
│     ├────────────────
│     │ null_resource.example is null
│
│ This value is null, so it does not have any attributes.
```

The only available data access patterns for possibly disabled resources are:

- `null_resource.example != null ? null_resource.example.id : "default value"`
- `try(null_resource.example.id, "default_value")`
- `can(null_resource.example.id) ? null_resource.example.id : "default value"`

> Notice this can be extended by using custom provider functions. Let's say if someone would like to

Reviewer 4: ```suggestion
Notice this can be extended by using custom provider-defined functions. Let's say if someone would like to
```

create a function that add a warning instead of raising an error when trying to access the attribute,
they could do:

- `provider::custom::warning_enabled(null_resource.example.id, "default value")

For now, we will be offering the three options above. 
`try` is the most concise option, but it should be used carefully since it can silently mask
different errors not caused by null values access. `can` and `!= null` are more verbose but 
are more reliable on doing what you're trying to do.

Discarded options:

> - In a GitHub thread, someone mentioned using `?` to deal with disabled resources. Since this would be a HCL change, we discarded that option.

Reviewer 4: Can you clarify why we dont want to introduce a HCL change please?

### What happens when it's disabled

Let's suppose we have created a resource using the `lifecycle -> enabled` field and then we want to disable it.
The behavior is going to be the same as if a resource is being destroyed:

```
# aws_instance.demo_vm_2 will be destroyed
# (because enabled is false)
  - resource "aws_instance" "demo_vm_2" {
      - ami                                  = "ami-07df274a488ca9195" -> null
      - arn                                  = "arn:aws:ec2:eu-central-1:532199187081:instance/i-0b63433033e61d818" -> null
      - associate_public_ip_address          = true -> null
      - availability_zone                    = "eu-central-1b" -> null
```

### What types of variables are supported on the conditional

There are a few types of values that cannot be used in this conditional:

1. Unknown values, since they cannot be used until the apply phase
2. Sensitive values
3. Null values
4. Conditionals that do not evaluate to true/bool values

## Future considerations

> 

Reviewer 4: Please add some considerations around supporting count+enabled/for-each+enabled.

## Current status

Issue: https://github.com/opentofu/opentofu/issues/1306

> [!NOTE]
> Every time we refer to *resources* in this RFC, we're talking about [managed resources](https://opentofu.org/docs/language/resources/), [data sources](https://opentofu.org/docs/language/data-sources/) and [ephemeral resources](https://github.com/opentofu/opentofu/issues/2834).

Right now, OpenTofu supports conditional enable/disable of resources by using `count` as a workaround:

```
variable "enabled" {
  default = true
}

resource "aws_instance" "example" {
  count = var.enabled ? 1 : 0
  # ...
}
```

This approach brings a few problems, such as adding indexes to resources that would be single instances, making it harder to manage using these indexes.

This RFC proposes a new way to do this in a cleaner and more semantic manner by using a new field on the `lifecycle` block called `enabled`:

```
resource "aws_instance" "example" {
  # ...

  lifecycle {
    enabled = var.enabled
  }
}
```

This is available not only for resources, but for modules too:

```
module "modcall" {
  source = "./mod"

  lifecycle {
    enabled = var.env == "production"
  }
}
```

## Proposed Solution

> 1. Support for conditional enabling on:

Reviewer 4: I think we still need to explicitly mention in the RFC that the default for this `enabled` field is `true`
